### PR TITLE
Added annealing and genetic algorithms for local search

### PR DESCRIPTION
**Pull request by rkruegs123**

### CODE REVIEW

File: pyzx/__init__.py [hunk 1]
```diff
@@ -26,6 +26,8 @@
 from .extract import *
 from .io import *
 from .tensor import *
+from .simulated_annealing import *
+from .genetic import *
```
Comment (jvdwetering) [inline pyzx/__init__.py]: I don't think we should import these into global scope

File: pyzx/__init__.py [hunk 2]
```diff
@@ -43,7 +45,9 @@
 from . import simulate
 from . import editor
 from . import routing
-from . import sparsify
+from . import scores
+from . import simulated_annealing
+from . import genetic
 from .routing.parity_maps import CNOT_tracker
```
Comment (jvdwetering) [inline pyzx/__init__.py:49]: Could we maybe put all of these under a single module, that might be a little less messy.

File: pyzx/congruences.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
+import numpy as np
```
Comment (jvdwetering) [inline pyzx/congruences.py]: Missing the standard header on top of the file. Also should have a top level documentation comment.

Comment (jvdwetering) [reply pyzx/congruences.py]: Also, there is no mypy annotation here, which I guess is fine for now, but I would like pyzx to have annotations at least for all the function signatures. All the functions are also missing documentation strings.

File: pyzx/genetic.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
+from tqdm import tqdm
```
Comment (jvdwetering) [inline pyzx/genetic.py]: Missing the standard header on top of the file. Also should have a top level documentation comment. Also, there are no documentation strings for all the classes.

File: pyzx/simplify.py [hunk 5]
```diff
@@ -356,3 +356,98 @@ def clifford_iter(g: BaseGraph[VT,ET]) -> Iterator[Tuple[BaseGraph[VT,ET],str]]:
     yield from pivot_iter(g)
     yield from id_iter(g)
     yield from spider_iter(g)
+
+
+def is_graph_like(g):
+    # checks that all spiders are Z-spiders
```
Comment (jvdwetering) [inline pyzx/simplify.py:364]: Should be a documentation string """using these quotes""" instead of a regular comment.

File: pyzx/simplify.py [hunk 6]
```diff
@@ -356,3 +356,98 @@ def clifford_iter(g: BaseGraph[VT,ET]) -> Iterator[Tuple[BaseGraph[VT,ET],str]]:
...
+    for z in zs:
+        b_neighbors = [n for n in g.neighbors(z) if g.type(n) == VertexType.BOUNDARY]
+        if len(b_neighbors) > 1:
+            return False
+
+    return True
+
+
+def to_graph_like(g):
+    # turn all red spiders into green spiders
```
Comment (jvdwetering) [inline pyzx/simplify.py:404]: Same as above

File: pyzx/simulated_annealing.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
+from tqdm import tqdm
```
Comment (jvdwetering) [inline pyzx/simulated_annealing.py]: Missing the standard header on top of the file. Also should have a top level documentation comment. Missing documentation strings for all the functions.

**Review by jvdwetering — COMMENTED** [review]
If you could add a bit of documentation to all the methods you've added that would be great. Mypy annotations for at least the function signatures would also be welcome, but I understand it is a bit of a pain. I think the three files you've added could also be part of the same submodule as to not further pollute the (already crowded) namespace. Also, please add a license statement to the top of each file.

**Comment by rkruegs123** [comment]
Sounds good. I've made these changes aside from the typing annotations. Happy to do this as well, but will take a bit longer.

**Comment by akissinger** [comment]
I think typing is not necessary for now. I'm eager to start doing some experiments with your code.

**Pull request merged**